pravega/pravega
# Issue 4653: Control rate of parallel Segment Container recoveries

**Change log description**  
Control the number of container starts/recoveries performed in parallel.

**Purpose of the change**  
Fixes #4653.

**What the code does**  
Until now, a Segment Store could start Segment Containers in parallel, which also includes the process of recovery in the case that the Segment Container to start is not new. However, if we think about a Segment Store recovering in parallel multiple Segment Containers, it may be the case that the data to process for each container is larger than the size of the cache (not to mention that the cache can still be used by other containers). This PR mainly sets a limit to the number of parallel Segment Container starts/recoveries to prevent overloading the cache.

To this end, this PR includes the following changes:

- In `ZKSegmentContainerMonitor`, we use a semaphore to control the number of parallel executions of `startContainer` method. The value of the semaphore can be parameterized.
- Added a new configuration parameter (`pravegaservice.container.parallelStarts`) to control the number of parallel container starts via configuration.

Note that this PR contains the same changes as in #4655, but replacing the use of `OrderedItemProcessor` by a semaphore to throttle the execution of segment container recoveries.

**How to verify it**  
All tests should pass as before this PR. Multiple system test runs have passed.
This is an example of the throttling of container recoveries in system tests:
```
(Without this PR)
segmentstore_2019-03-08.0.log.gz:2019-03-08 01:41:11,983 11410 [core-24] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[0] Recovery started.
segmentstore_2019-03-08.0.log.gz:2019-03-08 01:41:12,033 11460 [core-2] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[1] Recovery started.
segmentstore_2019-03-08.0.log.gz:2019-03-08 01:41:12,172 11599 [core-24] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[0] Recovery completed. Epoch = 1, Items Recovered = 0, Time = 189ms.
segmentstore_2019-03-08.0.log.gz:2019-03-08 01:41:12,174 11601 [core-2] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[1] Recovery completed. Epoch = 1, Items Recovered = 0, Time = 140ms.
segmentstore_2019-03-08.0.log.gz:2019-03-08 01:41:12,191 11618 [core-22] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[2] Recovery started.
segmentstore_2019-03-08.0.log.gz:2019-03-08 01:41:12,259 11686 [core-22] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[2] Recovery completed. Epoch = 1, Items Recovered = 0, Time = 67ms.
segmentstore_2019-03-08.0.log.gz:2019-03-08 01:41:12,333 11760 [core-2] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[3] Recovery started.
segmentstore_2019-03-08.0.log.gz:2019-03-08 01:41:12,361 11788 [core-2] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[3] Recovery completed. Epoch = 1, Items Recovered = 0, Time = 28ms.
```
```
(With this PR setting pravegaservice.container.parallelStarts=1)
segmentstore_2020-06-11.0.log.gz:2020-06-11 10:42:54,977 11263 [core-8] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[0] Recovery started.
segmentstore_2020-06-11.0.log.gz:2020-06-11 10:42:55,140 11426 [core-8] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[0] Recovery completed. Epoch = 1, Items Recovered = 0, Time = 162ms.
segmentstore_2020-06-11.0.log.gz:2020-06-11 10:42:55,297 11583 [core-7] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[1] Recovery started.
segmentstore_2020-06-11.0.log.gz:2020-06-11 10:42:55,312 11598 [core-7] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[1] Recovery completed. Epoch = 1, Items Recovered = 0, Time = 15ms.
segmentstore_2020-06-11.0.log.gz:2020-06-11 10:42:55,379 11665 [core-3] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[2] Recovery started.
segmentstore_2020-06-11.0.log.gz:2020-06-11 10:42:55,399 11685 [core-3] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[2] Recovery completed. Epoch = 1, Items Recovered = 0, Time = 20ms.
segmentstore_2020-06-11.0.log.gz:2020-06-11 10:42:55,447 11733 [core-20] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[3] Recovery started.
segmentstore_2020-06-11.0.log.gz:2020-06-11 10:42:55,454 11740 [core-20] INFO  i.p.s.server.logs.RecoveryProcessor - RecoveryProcessor[3] Recovery completed. Epoch = 1, Items Recovered = 0, Time = 7ms.
```
As you can see, without this PR multiple container start may happen concurrently. Instead, with this PR and setting `pravegaservice.container.parallelStarts=1` we force container starts to happen one at a time.
